lib/backend/backend.go

```diff
@@ -400,3 +400,55 @@ func NewLease(item Item) *Lease {
 		Revision: item.Revision,
 	}
 }
+
+// BatchPutter is an optional interface that backends can implement
+// to support batched PutBatch operations for improved performance when writing
+// multiple items at once.
+type BatchPutter interface {
+	// PutBatch upserts multiple items into the backend in a single call, in a way
+	// that is equivalent to a loop around multiple invocations of [backend.Put],
+	// but with the potential to be more efficient or faster, depending on the
+	// implementation. Returns a revision item for each item in the same order.
+	// Revisions are not guaranteed to be different nor they are guaranteed to be
+	// the same between items of the same batch. If an error is returned, it's
+	// possible for some of the items to have been persisted to the storage. The
+	// order in which items are internally persisted is an implementation detail.
+	PutBatch(context.Context, []Item) ([]string, error)
+}
+
+// PutBatch is an implementation of PutBatch that by default calls Put for each item.
+// Backends can overwrite this behavior providing optimized PutBatch implementation.
+//
+// WARNING: Make sure that items have unique keys when calling PutBatch to avoid fallback
+// to single Put calls.
+func PutBatch(ctx context.Context, bk Backend, items []Item) ([]string, error) {
+	// Many Backend implementations rely on unique keys for correct operation.
+	// Where it is up to the caller to ensure this to remove duplication keys
+	// Just in case we will fallback to single Put calls if duplicates are detected.
+	if v, ok := bk.(BatchPutter); ok && !hasDuplicateKeys(items) {
+		revs, err := v.PutBatch(ctx, items)
+		return revs, trace.Wrap(err)
+	}
```

Comment:
The point was not to trigger a potentially slow sequential Put loop, the point was to either deduplicate the items by key or make it an error - it's not like it makes sense to Put the same key twice, the event stream is not guaranteed to contain every change to every item.

Reply:
58d1e736ed270c261d5d93b8fbd09b063dbda698